compiler/rustc_const_eval/src/interpret/operand.rs

```diff
@@ -564,8 +564,16 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 throw_inval!(AlreadyReported(reported))
             }
             ty::ConstKind::Unevaluated(uv) => {
+                // NOTE: We evaluate to a `ValTree` here as a check to ensure
+                // we're working with valid constants, even though we never need it.
                 let instance = self.resolve(uv.def, uv.substs)?;
-                Ok(self.eval_to_allocation(GlobalId { instance, promoted: uv.promoted })?.into())
+                let cid = GlobalId { instance, promoted: None };
+                let _valtree = self
+                    .tcx
+                    .eval_to_valtree(self.param_env.and(cid))?
+                    .unwrap_or_else(|| bug!("unable to create ValTree for {:?}", uv));
+
```

Comment:
This is a strange check. Why would CTFE care whether the constant is a valid valtree? What if `uv` has a type that does not even allow valtree construction, like a raw pointer?

Reply:
> But then IMO it would make most sense to have a function that evaluates a ty::Const to a Valtree, and always go through that. 

I will do that in https://github.com/rust-lang/rust/pull/104317